src/ir/auth_logic/ast.h

```diff
@@ -0,0 +1,182 @@
+        std::string name,
+        std::vector<std::string> args,
+        AstSign sign)
+      : name_(std::move(name)),
+        args_(std::move(args)),
+        sign_(std::move(sign)) {}
+  private:
+    std::string name_;
+    std::vector<std::string> args_;
+    AstSign sign_;
+};
+
+// This represents an expression that applies either a predicate or a 
+// "canActAs" to a principal, and it is a sub-term of AstFlatFact
+//    - pred_name(arg_1, ..., arg_n)           (a predicate)
+//    - principal_1 canActAs principal_2       (a canActAs)
+// To encapsulate which form this takes, it:
+//    - uses factory methods to make either form
+//    - uses a private std::variant member that takes either of these two forms
+//    - uses std::visit to do operations on the actual state
+class AstVerbPhrase {
+  public:
+    std::unique_ptr<AstVerbPhrase>
+      CreateVerbPhrase(AstPredicate pred);
+    std::unique_ptr<AstVerbPhrase>
+      CreateVerbPhrase(AstPrincipal right_prin);
+  private:
+    // The principal variant of the internal form means this represents
+    // "canActAs principal".
+    using InternalFormType = std::variant<AstPredicate, AstPrincipal>;
+    InternalFormType internal_form_;
+    AstVerbPhrase(InternalFormType internal_form)
+      : internal_form_(std::move(internal_form)) {}
+};
+
+// This the AST Node for FlatFact which takes on one of two forms:
+//   - <principal> <verb_phrase>  (* a principal applied to a verbphrase *)
+//   - <predicate>                (* just a predicate *)
+// A similar variant-based pattern is used for this as for VerbPhrase. Here an 
+// inner struct is used for the first form (principal applied to verb_phrase).
+class AstFlatFact {
+  public:
+    std::unique_ptr<AstFlatFact>
+      CreateFlatFact(AstPrincipal prin, AstVerbPhrase verb_phrase);
+    std::unique_ptr<AstFlatFact>
+      CreateFlatFact(AstPredicate prin);
+  private:
+    typedef struct VerbPhraseFlatFact {
+      AstPrincipal prin;
+      AstVerbPhrase verb_phrase;
+      VerbPhraseFlatFact(AstPrincipal prin, AstVerbPhrase verb_phrase)
+        : prin(std::move(prin)),
+          verb_phrase(std::move(verb_phrase)) {}
+    } VerbPhraseFlatFact;
+    using InternalFormType = std::variant<VerbPhraseFlatFact, AstPredicate>;
+    InternalFormType internal_form_;
+    AstFlatFact(InternalFormType internal_form)
+        : internal_form_(std::move(internal_form)) {}
+};
+
+// This is the AST node for facts which have two forms:
+//  - <flat_fact>
+//  - <principal> canSay <flat_fact>
+class AstFact {
+  public:
+    std::unique_ptr<AstFact> CreateFact(AstFlatFact flat_fact);
+    std::unique_ptr<AstFact> CreateFact(AstPrincipal prin, AstFact* fact);
+  private:
+    typedef struct CanSayFact {
+      AstPrincipal prin;
+      AstFact *fact;
+      CanSayFact(AstPrincipal prin, AstFact* fact)
+        : prin(std::move(prin)),
+          fact(fact) {}
+    } CanSayFact;
+    using InternalFormType = std::variant<AstFlatFact, CanSayFact>;
```

Comment:
[Internal alias in interface](https://google.github.io/styleguide/cppguide.html#Aliases), as mentioned elsewhere.

Reply:
done.